### PR TITLE
allow compatibility() for lockfiles

### LINKED ISSUE #12773: [question] locking with compatibility
### What is your question?

I want to use `compatibility`  in combination with locking and somehow it does not work (conan 1.54).

I followed the approach off address sanitizer https://docs.conan.io/en/1.56/howtos/sanitizers.html?highlight=address , i.e. 
adding a sanitizer key to the `compiler` settings and making a special profile for address sanitizer:

```
compiler.sanitizer=Address
build_type=Debug
[options]
[build_requires]
[env]
CFLAGS="/fsanitize=address -D_DISABLE_VECTOR_ANNOTATION"
CXXFLAGS="/fsanitize=address -D_DISABLE_VECTOR_ANNOTATION"
```

I added in the recipes a `compatibility` part to make this profile compatible with the normal debug profile:

```
    def compatibility(self):
        # make pkg without sanitizer compatible
        if ( self.settings.compiler == "msvc" and self.settings.compiler.sanitizer == "Address"):
           return [ {"settings": [("compiler.sanitizer", "None")]}, ]
```

I used the example the example https://github.com/conan-io/examples.git features/editable/cmake and modified the say package adding the above compatible section. 

If I create the package say for debug without sanitizer enabled, I can create the hello package and it indeed does the compatibility part and it is created:

```
say/0.1: Main binary package '16cfe13190cf7321cc7f619b134215b5d133f553' missing. Using compatible package '13542e3278eff2b7de4be1519f98572409d630c3'
conanfile.txt: Installing package
Requirements
    say/0.1 from local cache - Cache
Packages
    say/0.1:13542e3278eff2b7de4be1519f98572409d630c3 - Cache
```

But now I try to combine with with locking first:

First the base lock (I use profiles here already to be sure that this is not causing the issue:
```
λ \venv\Scripts\conan lock create conanfile.txt --lockfile-out=locks/base.lock --base -pr:h vs2022_x64_profile_debug_asan -pr:b vs2022_x64_profile
say/0.1: Main binary package '16cfe13190cf7321cc7f619b134215b5d133f553' missing. Using compatible package '13542e3278eff2b7de4be1519f98572409d630c3'
Requirements
    say/0.1 from local cache - Cache
Packages
    say/0.1:13542e3278eff2b7de4be1519f98572409d630c3 - Cache

Generated lockfile: D:\dev\conan\examples2\features\editable\cmake\hello\locks\base.lock
```

After that the profile lock using the base lock:
```
 \venv\Scripts\conan lock create conanfile.txt --lockfile=locks/base.lock --lockfile-out=locks/profile.lock  -pr:h vs2022_x64_profile_debug_asan -pr:b vs2022_x64_profile
say/0.1: Main binary package '16cfe13190cf7321cc7f619b134215b5d133f553' missing. Using compatible package '13542e3278eff2b7de4be1519f98572409d630c3'
Requirements
    say/0.1 from local cache - Cache
Packages
    say/0.1:13542e3278eff2b7de4be1519f98572409d630c3 - Cache

Generated lockfile: D:\dev\conan\examples2\features\editable\cmake\hello\locks\profile.lock
```

And after that the conan install using the lock:
```
λ \venv\Scripts\conan install . --lockfile=locks/profile.lock
Using lockfile: 'D:\dev\conan\examples2\features\editable\cmake\hello\locks\profile.lock'
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=msvc
compiler.cppstd=20
compiler.runtime=static
compiler.runtime_type=Debug
compiler.sanitizer=Address
compiler.version=193
os=Windows
os_build=Windows
[options]
[build_requires]
[env]
CFLAGS=/fsanitize=address -D_DISABLE_VECTOR_ANNOTATION
CXXFLAGS=/fsanitize=address -D_DISABLE_VECTOR_ANNOTATION
[conf]
tools.cmake.cmaketoolchain:generator=Ninja
tools.cmake.cmaketoolchain.presets:max_schema_version=3
tools.env.virtualenv:auto_use=True
[buildenv]
CTEST_OUTPUT_ON_FAILURE=1

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=static
compiler.runtime_type=Release
compiler.sanitizer=None
compiler.version=193
os=Windows
os_build=Windows
[options]
[build_requires]
[env]
[conf]
tools.cmake.cmaketoolchain:generator=Ninja
tools.cmake.cmaketoolchain.presets:max_schema_version=3
tools.env.virtualenv:auto_use=True
[buildenv]
CTEST_OUTPUT_ON_FAILURE=1

ERROR: 'say/0.1#2ef6637a5c616c73da7208377074d010' package-id '16cfe13190cf7321cc7f619b134215b5d133f553' doesn't match the locked one '13542e3278eff2b7de4be1519f98572409d630c3'
```

Is this a bug, or am I doing something wrong? The above works fine without the compatibility part, i.e. creating the ASAN package for say.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Bugfix: Allow ``compatibility()`` for lockfiles.
Docs: Omit

Close https://github.com/conan-io/conan/issues/12773

### CODE REVIEW

**Review by AbrilRBS — APPROVED** [review]

**Comment by SSE4** [comment]
this is a bit weird. lockfile is supposed to lock the package, means always use the package being locked during the lockfile creation. it shouldn't suddenly fallback to a compatible package during the graph resolution. am I missing the point?

**Comment by memsharded** [comment]
> this is a bit weird. lockfile is supposed to lock the package, means always use the package being locked during the lockfile creation. it shouldn't suddenly fallback to a compatible package during the graph resolution. am I missing the point?

It is now (with the fix) locking the package, it was failing to resolve to the locked package, not to fallback to a compatible one.

**Comment by SSE4** [comment]
> It is now (with the fix) locking the package, it was failing to resolve to the locked package, not to fallback to a compatible one.

okay, good to know. maybe rephrase title/description to indicate that clearly. right now, it reads like `compatibility()` is added to the lockfile, like lockfiles now uses compatible packages during the resolution.

**Review by czoido — APPROVED** [review]
